flow/layers/opacity_layer.cc

```diff
@@ -12,6 +12,20 @@ namespace flutter {
 OpacityLayer::OpacityLayer(SkAlpha alpha, const SkPoint& offset)
     : alpha_(alpha), offset_(offset) {}
 
+void OpacityLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  DiffContext::AutoSubtreeRestore subtree(context);
+  auto* prev = reinterpret_cast<const OpacityLayer*>(old_layer);
+  if (!context->IsSubtreeDirty()) {
+    assert(prev);
+    if (alpha_ != prev->alpha_ || offset_ != prev->offset_) {
+      context->MarkSubtreeDirty(old_layer->paint_region());
+    }
+  }
+  context->PushTransform(SkMatrix::Translate(offset_.fX, offset_.fY));
+  DiffChildren(context, prev);
```

Comment:
In those layers that subclass MergedContainerLayer, we might want to watch which children we diff. They have a direct child that differs on just about every frame because that child is synthetic. The real children of a MergedContainerLayer are the children of the synthetic child.

See MergedContainerLayer::GetChildContainer().

Reply:
I see. I think `DiffChildren` can be made virtual and then call `GetChildContainer()->DiffChildren(context, old_layer->GetChildContainer())`; That way we skip the synthetic child and just compare actual children.

Reply:
Another solution might be to copy the original id of the child container across. So, for MCLs they copy `this->original_id = prev->original_id` as usual followed by `GetChildContainer()->original_id = prev-GetChildContainer()->original_id`?

They don't currently do anything specific to skip the extra child container on Preroll and Paint, but there is no compelling reason for them to do that other than to save a couple of method calls (though, now that I think about it, it might save an extra bounds computation and assignment in the Preroll stage, still it would leave the tree with an incompletely initialized ContainerLayer below each MCL).

Skipping the child container in the Diff would require every implementation of Diff to special case the call to Diff the children, but copying the ID across might achieve the same result without having to special case anything...?

Reply:
The `DiffChildren` would be just implemented on `MergedContainerLayer`, but I think if the ChildContainer preserves the `original_layer_id` of previous ChildContainer as you suggest the effect will be same. And it does seem cleaner that way because it doesn't need to "skip" part of hierarchy.

Reply:
Should be fixed [here](https://github.com/flutter/engine/pull/21498/commits/5f52aff6939108f755f3631981a76665b98b8bde#diff-fc0e3d96b0f5ffcb2ca12dd80cf73019R221).